go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -0,0 +1,489 @@
+	c.handleErr(err, eKey)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	ns, name, keyErr := cache.SplitMetaNamespaceKey(key.(string))
+	if keyErr != nil {
+		klog.ErrorS(err, "Failed to split meta namespace cache key", "key", key)
+	}
+	metrics.MetricRequeueServiceCount.WithLabelValues(key.(string)).Inc()
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		klog.V(2).InfoS("Error syncing service, retrying", "service", klog.KRef(ns, name), "err", err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping service %q out of the queue: %v", key, err)
+	c.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (c *Controller) syncServices(key string) error {
+	startTime := time.Now()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Processing sync for service %s on namespace %s ", name, namespace)
+	metrics.MetricSyncServiceCount.WithLabelValues(key).Inc()
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing service %s on namespace %s : %v", name, namespace, time.Since(startTime))
+		metrics.MetricSyncServiceLatency.WithLabelValues(key).Observe(time.Since(startTime).Seconds())
+	}()
+
+	service, err := c.serviceLister.Services(namespace).Get(name)
+	// Since we're getting stuff from a local cache, it is basically impossible to get this error.
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// Delete the Service from OVN if:
+	// - it was deleted from the cache (doesn't exist in Kubernetes anymore)
+	// - it mutated to a new service Type that we don't handle (ExternalName, Headless)
+	if err != nil || !util.ServiceTypeHasClusterIP(service) || !util.IsClusterIPSet(service) {
+		klog.Infof("Deleting service %s on namespace %s from OVN", name, namespace)
+		// Obtain the VIPs associated to the Service from the Service Tracker
+		vips := c.serviceTracker.getService(name, namespace)
+		for vipKey := range vips {
+			// the VIP is stored with the format IP:Port/Protocol
+			vip, proto := splitVirtualIPKey(vipKey)
+			// Get the OVN LoadBalancer
+			lbID, err := loadbalancer.GetOVNKubeLoadBalancer(proto)
+			if err != nil {
+				klog.Errorf("Error getting OVN LoadBalancer for protocol %s", proto)
+				return err
+			}
+			// Delete the Service VIP from OVN
+			if err := loadbalancer.DeleteLoadBalancerVIP(lbID, vip); err != nil {
+				klog.Errorf("Error deleting VIP %s on OVN LoadBalancer %s", vip, lbID)
+				return err
+			}
+			// Delete the Service VIP from the Service Tracker
+			c.serviceTracker.deleteServiceVIP(name, namespace, vip, proto)
+		}
+		// Delete the Service form the Service Tracker
+		c.serviceTracker.deleteService(name, namespace)
+		return nil
+	}
+	klog.Infof("Creating service %s on namespace %s on OVN", name, namespace)
+	// The Service exists in the cache: update it in OVN
+	// Get the endpoint slices associated to the Service
```

Comment:
This is one of the reasons why I choose to implement it this way, IIRC that happens becuase kube-proxy is using a tracker for the endpoints, the tracker is being updated with the endpointslices events
With the current approach, once we have an event on an slice or a service, we go to the cache, obtain ALL the slices that belong to the Service and we create the enpoint list with the things we found in the cache, this is one of the parts that I need more feedback about, theoretically I think that it solves the problem of working with events

Reply:
> Hm... so there was that endpointslice race condition where if an endpoint X moves from slice A to slice B, and you get the update event for slice A and then process it before slice B is updated, then you temporarily lose endpoint X. This controller does not yet attempt to deal with that, correct?

I ´ve talked with RobScott, and it seems that most of the problems they have in kubeproxy with slices comes of using the EndpointMap structure:
> The issue we had with kube-proxy was that we were storing data in a different data structure (EndpointsMap) and simply adding/removing endpoints to that as EndpointSlice events came in would not be sufficient since events could come in out of order

He´s going to evaluate this approach of using an informer. We don´ t deal with events to update the endpoints map here, each update reconcile the endpoints from the existing slices in the cache (and deduplicate them see ` etLbEndpoints(endpointSlices, svcPort, family)` , we can´ t have that situation ... but there can be other of course :)

Reply:
I don't see how it matters what data structures you use. There's no guarantee that you'll be notified about the updates to Slice A and Slice B close enough together that you'll process them in the same resync. ie:

- EndpointSliceController updates Slice A (removing endpoint X), and then Slice B (adding endpoint X)
- kube-proxy/ovn-kubernetes sees the Update event for Slice A, queues a resync
- (If you want, you can imagine that there's network trouble at this point, and the apiserver's first attempt to notify about the Update to Slice B gets dropped. But that's not really necessary.)
- kube-proxy/ovn-kubernetes runs its queued resync, using the data from the informer cache, which contains the old state of B and the new state of A. Endpoint X gets dropped.
- kube-proxy/ovn-kubernetes sees the Update event for Slice B, queues a resync
- kube-proxy/ovn-kubernetes runs its queued resync, Endpoint X gets re-added.

Reply:
> I don't see how it matters what data structures you use. There's no guarantee that you'll be notified about the updates to Slice A and Slice B close enough together that you'll process them in the same resync. ie:

is not the data structure only, is how you process the updates, with the endpoint map you never recover because you are processing only updates and if they come out of order they screw you,

> kube-proxy/ovn-kubernetes runs its queued resync, Endpoint X gets re-added.

because you always resyc,  you' ll be "eventually consistent" , there is no magic, you need to live with that, the important is to be able to recover fast and you depend on the apiserver here to tell you current state, and he is allowed to tell you out of order

Reply:
> because you always resyc, you' ll be "eventually consistent"

My point is that with Endpoints, every listener has the invariant "for every service, at any given time, either (1) you haven't learned anything about its endpoints yet, or (2) you know a set of endpoints that were previously 100% correct, or (3) you know a set of endpoints that are currently 100% correct". With EndpointSlice, the invariant is "for every service, either (1) you haven't learned anything about its endpoints yet, or (2) you know a set of endpoints which may or may not be complete, and may or may not be internally self-consistent".

> there is no magic, you need to live with that

You don't need magic, you just need more information from the EndpointSliceController to understand when a slice is standalone and when it's part of a set...

Reply:
> You don't need magic, you just need more information from the EndpointSliceController to understand when a slice is standalone and when it's part of a set...

ahh, yeah, you are right ... but I think that this discussion applies here
https://github.com/kubernetes/kubernetes/pull/93520#issuecomment-671944119